### DISCUSSION TITLE
TypeError when running examples

### DISCUSSION THREAD

**Discussion by VerwimpEli** [ContinualAI/avalanche#515] [Q&A] [answered]
Hi,

Upon running the examples I run into a runtime TypeError:

in `dataset_utils.py`:
`   File "replay.py", line 31, in <module>
    from avalanche.benchmarks import nc_scenario`
 ...
`    class SubsetWithTargets(Generic[T_co, TTargetType], Subset[T_co]):
TypeError: 'type' object is not subscriptable`

It's the generic `Subset` that's causing problems, I believe. I'm running in a clean conda environment, python 3.8.

Thanks!

**Comment by vlomonaco** [comment, marked as answer]
Thanks @VerwimpEli for reporting this. @lrzpellegrini will follow-up on this soon! In the meanwhile, can you provide a minimal script to reproduce the issue?

**Reply by VerwimpEli** [reply]
`import avalanche.benchmarks.utils` already fails. So I assume it must be a problem from my side, since I don't see anybody else running in this problem.

I've created my conda environment as mentioned in the developer instructions on the continualai [site](https://avalanche.continualai.org/getting-started/how-to-install#developer-mode-install). Using python 3.8, pytorch 1.8.1, running on a linux (Ubuntu 20.4) machine.

It looks a lot like the Mypy runtime problems mentioned [here](https://mypy.readthedocs.io/en/stable/runtime_troubles.html), but adding the future import `from __future__ import annotations` did not resolve the problem.

Removing the generic typing (i.e. `Subset` rather than `Subset[T_co]`) solves the problem for this line, but there are more lines with the same issue. 

(Possibly) relevant package versions:
`typing_extensions         3.7.4.3`

**Reply by VerwimpEli** [reply]
Fixed by removing the indexing on the second types.

 on line 63 in `avalanche_dataset.py`:
`class AvalancheDataset(IDatasetWithTargets[T_co, TTargetType], Dataset[T_co]):` 
changes to
`class AvalancheDataset(IDatasetWithTargets[T_co, TTargetType], Dataset):` 

and on line 173 in `dataset_utils.py`:
`class SubsetWithTargets(Generic[T_co, TTargetType], Subset[T_co]):`
changes to 
`class SubsetWithTargets(Generic[T_co, TTargetType], Subset):`

Still not sure whether this is a bug (in avalanche, or somewhere else) or a wrong configuration from my part.

**Reply by lrzpellegrini** [reply]
Hi @VerwimpEli, I'm trying to reproduce the error. For the moment I'm installing a fresh conda environment by following the "developer mode" guide. I'll need those info to look into this issue:

- Which version of Avalanche you are using (the day you cloned the repo)
- The list of installed packages in your conda environment (you can use the `conda list` command)
- Possibly, the exact steps you used when creating the environment

This will help me a lot, thank you!

**Reply by VerwimpEli** [reply]
Hi,

- I cloned the repo yeseterday (9/04)
- [conda_list.txt](https://github.com/ContinualAI/avalanche/files/6285290/conda_list.txt)
- Other than installing pytorch and adding the avalanche path to my pythonpath, I did nothing

Thanks for looking into this!

**Reply by lrzpellegrini** [reply]
Hi, I can't reproduce the issue and the environment configuration. In order to create the environment did you use the bash script or did you follow a manual procedure?

**Reply by VerwimpEli** [reply]
Initially I installed everything manually like the instructions mentioned, but the bash script results in the same errors...

**Reply by vlomonaco** [reply]
@VerwimpEli I'm also unable to replicate the issue ☹️ Can you try re-creating an environment from scratch maybe using py3.9?

```
# Step 1
git clone https://github.com/vlomonaco/avalanche.git
cd avalanche
conda create -n avalanche-env python=3.9 -c conda-forge
conda activate avalanche-env

# Step 2
# Istall Pytorch with Conda (instructions here: https://pytorch.org/)

# Step 3
conda env update --file environment.yml
```

Let us know how it goes!

**Reply by VerwimpEli** [reply]
Using python 3.9 the issue is solved. Maybe a bug in the specific python version I had? Anyhow, glad it's working properly now.

**Reply by vlomonaco** [reply]
Thanks for you patience @VerwimpEli! Marking this as solved! :)

**Reply by GengDavid** [reply]
Hi, I also met this problem with python 3.7.7, pytorch version 1.5.1
Would there be any solution rather than updating the python version?

**Reply by vlomonaco** [reply]
mmh maybe you can try update the "typing_extensions" package? what's your version?

**Reply by GengDavid** [reply]
I have tried typing_extensions==3.7.4.3 / 3.10.0.0 / 3.10.0.1 / 3.10.0.2. The problem remains. 😢

**Reply by GengDavid** [reply]
Updating Python and PyTorch versions do solve the problem.

**Reply by SinHanYang** [reply]
Having the same issue when using python 3.7, torch 1.6.0, but fixed when using torch 1.12.0.
